### DISCUSSION TITLE
Copied Interface

### DISCUSSION THREAD

**Discussion by HemyO** [dotnet/csharplang#7092] [Language Ideas] [answered]
## Background

Sometimes when developing in C#, you may encounter this situation:

- You need to use two data that implement the same interfaces, but there may be multiple implementations of that data, such as in-game damage calculation, which may require 2 parameters related to numbers, such as: `short`, `int`, `long`, `float`, `double`. But, sometimes for compatibility with multiple types of numbers, it may need to duplicate these contents:

```csharp
// This may be a bad example because if this method is only for handling `Convert`, it may only constrain `X` and `Y` to implement `IConvertible`.
public static X CalcDamage<X, Y>(X weaponDamage, Y multiplier)
  where X : IComparable, IComparable<X>, IConvertible, IEquatable<X>, IFormattable
  where Y : IComparable, IComparable<X>, IConvertible, IEquatable<X>, IFormattable
  => // do something here ...

// client
(int a, double b) = (5, 5.5);
CalcDamage(a, b);
```

- You need to mimic the implementation of interfaces of types outside the projects, such as `int` in C# to implement the corresponding interfaces.

```csharp
// Here using .NET 6 as an example, there are too many interfaces in .NET 7.
// Every interfaces here are consistent with C# int except ISpanFormattable.
public readonly struct FantasyMoney : IComparable, IComparable<FantasyMoney>, IConvertible, IEquatable<FantasyMoney> {
  // define the data ...
}

public readonly struct FantasyFame : IComparable, IComparable<FantasyFame>, IConvertible, IEquatable<FantasyFame> {
  // define the data ...
}
```

After the opening of static virtual members in interfaces in .NET 7, this problem occur more frequently.

## Proposal

Let's define these 2 keywords: `as` and `remove`.

- `as`: It can define all the implementations or inherited interfaces of the types or interfaces to be used. For example, `A as int` will copy the implemented interfaces of `int` to `A` (`A` may be a type or interface). `int` has generic interfaces: `IComparable<int>` and `IEquatable<int>`. When copying, `int` will automatically be replaced with `A`. But, if the type or interface being used contains one or more type parameters, the type being transferred must also provide an equal number of parameters. For example, `B<X> as List<X>`, because `List` also needs an `X` type parameter to use, so `B` will also need a type parameter to use.

```csharp
// In this feature
A as int

B<X> as List<X>


// After transform
A : IComparable, IComparable<A>, IConvertible, IEquatable<A>, ISpanFormattable

B<X> : ICollection<X>, IEnumerable<X>, IList<X>, IReadOnlyCollection<X>, IReadOnlyList<X>, ICollection, IEnumerable, IList
```

- `remove`: If the interfaces copied by `as` are in the interfaces provided by `remove`, the duplicated interfaces will not be copied. For example, `A as int remove IEquatable` will remove `IEquatable` from `A`. If the removed interfaces are not in `int`, a warning will be displayed. Note! The interface after `remove` does not need to mark type parameter.

```csharp
// In this feature
A as int remove IEquatable

// After transform
A : IComparable, IComparable<A>, IConvertible, ISpanFormattable
```

## Example with this feature

Through `as` and `remove`, the previous situation can be changed to the following.

- In method:

```csharp
// This may be a bad example because if this method is only for handling `Convert`, it may only constrain `X` and `Y` to implement `IConvertible`.
public static X CalcDamage<X, Y>(X weaponDamage, Y multiplier)
  where X as int
  where Y as int
  => // do something here ...

// client
(int a, double b) = (5, 5.5);
CalcDamage(a, b);
```

- In type:

```csharp
// Here using .NET 6 as an example, there are too many interfaces in .NET 7
// Every interfaces here are consistent with C# int except ISpanFormattable.
public readonly struct FantasyMoney as int remove ISpanFormattable {
  // define the data ...
}

public readonly struct FantasyFame as int remove ISpanFormattable {
  // define the data ...
}
```

## Feature Problems

- This feature may decrease the readability of code, because it is necessary to understand what the type or interface after `as` is. But I believe it can be a choice between readability and convenience.

Thanks for your watching. 😊

**Comment by HaloFour** [comment, marked as answer]
Sounds like something that could be implemented via tooling, like an analyzer that could apply a fixer that would copy the interfaces from another type following the policies suggested above, or a source generator that could do the same.

**Reply by HemyO** [reply]
You're right. It's possible to do it with tooling. Also, it can implemented via attribute. Like this:
```csharp
[As<int>][Remove<ISpanFormattable>]
public readonly struct FantasyMoney {
  // define the data ...
}
```
If a problem can be solved in an existing tool or feature(like tool or attribute), it seems less necessary to add this.

Thanks for your answer. 😊

**Reply by huoyaoyuan** [reply]
It sounds a bit complex for the "transforming" logic of interfaces.
Such feature is more suitable for custom source generator. A language feature needs to cover more scenarios as possible and can be unnecessarily complex. Custom generator can implement the features you need.
